libkbfs/folder_branch_ops.go

```diff
@@ -738,6 +764,109 @@ func (fbo *folderBranchOps) startMonitorChat(tlfName tlf.CanonicalName) {
 	})
 }
 
+func (fbo *folderBranchOps) kickOffRootBlockFetch(
+	ctx context.Context, rmd ImmutableRootMetadata) <-chan error {
+	ptr := rmd.Data().Dir.BlockPointer
+	return fbo.config.BlockOps().BlockRetriever().Request(
+		ctx, defaultOnDemandRequestPriority, rmd, ptr, NewDirBlock(),
+		TransientEntry)
+}
+
+func (fbo *folderBranchOps) waitForRootBlockFetch(
+	ctx context.Context, rmd ImmutableRootMetadata,
+	fetchChan <-chan error) error {
+	fbo.log.CDebugf(ctx, "Ensuring that the latest root directory "+
+		"block for revision %d is available", rmd.Revision())
+	select {
+	case err := <-fetchChan:
+		fbo.log.CDebugf(ctx, "Root block fetch complete: +%v", err)
+		return err
+	case <-ctx.Done():
+		return errors.WithStack(ctx.Err())
+	}
+}
+
+func (fbo *folderBranchOps) commitFlushedMD(
+	rmd ImmutableRootMetadata, updatedCh <-chan struct{}) {
+	diskMDCache := fbo.config.DiskMDCache()
+	if diskMDCache == nil {
+		return
+	}
+
+	// Bail out if the latest merged revision has already been updated.
+	select {
+	case <-updatedCh:
+		return
+	default:
+	}
+
+	ctx := fbo.ctxWithFBOID(context.Background())
+	rev := rmd.Revision()
+	if fbo.config.IsSyncedTlf(fbo.id()) {
+		// For synced TLFs, wait for prefetching to complete for
+		// `rootPtr`. When it's successfully done, commit the
+		// corresponding MD.
+		rootPtr := rmd.Data().Dir.BlockPointer
+		fbo.log.CDebugf(ctx, "Fetching root block of revision %d, ptr %v",
+			rev, rootPtr)
+		rootCh := fbo.kickOffRootBlockFetch(ctx, rmd)
+		select {
+		case err := <-rootCh:
+			if err != nil {
+				fbo.log.CDebugf(ctx, "Error getting root block: %+v", err)
+				return
+			}
+		case <-updatedCh:
+			fbo.log.CDebugf(ctx, "The latest merged rev has been updated")
+			return
+		case <-fbo.shutdownChan:
+			fbo.log.CDebugf(ctx, "Shutdown, canceling root block wait")
+			return
+		}
+
+		fbo.log.CDebugf(ctx, "Waiting for prefetch of revision %d, ptr %v",
+			rev, rootPtr)
+		waitCh, err := fbo.config.BlockOps().Prefetcher().
+			WaitChannelForBlockPrefetch(ctx, rootPtr)
+		if err != nil {
+			fbo.log.CDebugf(ctx,
+				"Error getting wait channel for prefetch: %+v", err)
+			return
+		}
+
+		select {
+		case <-waitCh:
+		case <-updatedCh:
+			fbo.log.CDebugf(ctx, "The latest merged rev has been updated")
+			fbo.config.BlockOps().Prefetcher().CancelPrefetch(rootPtr.ID)
+			return
+		case <-fbo.shutdownChan:
+			fbo.log.CDebugf(ctx, "Shutdown, canceling prefetch wait")
+			return
+		}
```

Comment:
> This will take a loooooooong time for big TLFs, especially the first time. Although I guess it's always executed in a goroutine, and will be canceled any time an update comes in.

I think that's ok, since we'll short-circuit the wait if the user actually uses the MD to satisfy a read.  And until that happens, it's like the TLF doesn't exist for the user on that device, anyway.

> Would be good to think about ways to communicate progress. Maybe the prefetch tree could also sum byte counts of children instead of just number of blocks.

We can sort of know what the expected number of total (encoded/encrypted) bytes are, based on the total in the MD object.  But it might be tricky to keep track of that when prefetches are pending and new updated MDs come in.

This is a good idea if we can work it out, but should probably be in a separate PR and maybe won't be needed until there's a GUI design.  Want to file a ticket?

Reply:
> But it might be tricky to keep track of that when prefetches are pending and new updated MDs come in.

If we store the sums in the tree just like the block counts, I don't think it'll be any trickier to manage.

https://keybase.atlassian.net/browse/KBFS-3575